src/sage/categories/finite_dimensional_lie_algebras_with_basis.py

```diff
@@ -1187,74 +1196,184 @@ def sgn(k, X):
                         return zero, None
                     if k > val:
                         Y.insert(i+1, k)
-                        return mone**(i+1), tuple(Y)
+                        return (-1)**(i+1), tuple(Y)
                 Y.insert(0, k)
-                return R.one(), tuple(Y)
+                return 1, tuple(Y)
+
+            def HomAlgebra(Module):
+                """
+                Define the endomorphism Lie algebra of M, with generators {gi}.
+                """
+
+                n = dim(Module)
+                MS=MatrixSpace(R, n)
+                keys=[]
+                values=[]
+                gln=lie_algebras.gl(R, n)
+        
+                family = gln.structure_coefficients()
+        
+                for x in family.keys():
+                    i=gln.basis().keys().index(x[0])
+                    j=gln.basis().keys().index(x[1])
+                    keys.append((f'g{i}', f'g{j}'))
+        
+                for x in family.values():
+                    D = x.monomial_coefficients()
+                    B=list(D.keys())
+                    C=list(D.values())
+                    bracket = {f'g{gln.basis().keys().index(B[i])}': C[i] for i in range(len(D))}
+        
+                    values.append(bracket)
+        
+                d = {keys[i]: values[i] for i in range(len(keys))}
+        
+                str = ','.join(f'g{i}' for i in range(n^2))
+        
+                A=LieAlgebra(R, str, d)
+        
+                return A
+
+            def Lieaction(self, M, h, LieElement, ModuleElement):
+
+                n=dim(M)
+                G=HomAlgebra(M)
+                preimage=list(h.keys())
+        
+                image=list(h.values())
+        
+                gln=lie_algebras.gl(R, n)
+                values=[]
+                for x in image:
+                    D=x.monomial_coefficients()
+                    B=list(D.keys())
+                    C=list(D.values())
+                    output=0
+                    for i in range(len(B)):
+                        l=n*B[i][0]+B[i][1]
+                        output+=G.basis().values()[l]*C[i]
+                    values.append(output)
+        
+        
+                d = {preimage[i]:values[i] for i in range(len(preimage))}
+        
+                g=self.morphism(d)
+        
+                H=zero_matrix(R, n)
+                K=list(itertools.chain(*H))
+        
+                final = g(LieElement).monomial_coefficients()
+                F=list(final.keys())
+                N=list(final.values())
+        
+                for i in range(len(F)):
+                    l=G.basis().keys().index(F[i])
+                    K[l]+=N[i]
+        
+                MS=MatrixSpace(R, n)
+        
+                return (vector(ModuleElement))*MS(K)
 
             from sage.parallel.decorate import parallel
 
             @parallel(ncpus=ncpus)
-            def compute_diff(k):
+            def f1(k):
```

Comment:
$f1(k)$ is supposed to compute the first term of the differential map ($ \sum_{i=1}^p (-1)^{i+1}
                  (m g_i) \otimes g_1 \wedge \cdots \wedge
                  \hat{g}_i \wedge \cdots \wedge g_p$), and $f2(k)$ the second ($ \sum_{1 \leq i < j \leq p} (-1)^{i+j}
                  m \otimes [g_i, g_j] \wedge
                  g_1 \wedge \cdots \wedge \hat{g}_i
                  \wedge \cdots \wedge \hat{g}_j
                  \wedge \cdots \wedge g_p$). I will make the second part in parallel as well; the reason I split it into two was because I thought it made the code significantly more easy to navigate on the whole. Maybe I can just add a more detailed description besides each function?

Reply:
$f1(k)$ is supposed to compute the first term of the differential map ( $\sum (-1)^{i+1} (m g_i) \otimes g_1 \wedge \cdots \wedge \hat{g}i \wedge \cdots \wedge g_p$), and $f2(k)$ the second ( $\sum (-1)^{i+j} m \otimes [g_i, g_j] \wedge g_1 \wedge \cdots \wedge \hat{g}_i \wedge \cdots \wedge \hat{g}_j\wedge \cdots \wedge g_p$). I will make the second part in parallel as well; the reason I split it into two was because I thought it made the code significantly more easy to navigate on the whole. Maybe I can just add a more detailed description besides each function?

Reply:
I find it more complicated; in part because the names are not clear. Adding documentation will help, but I don't see a good reason to increase the complexity (by indirections and less code locality), which I find significantly harder to read and understand. You also run into parallel functions calling parallel functions, which is worrisome for me.

Reply:
OK, done.